presto-main/src/main/java/io/prestosql/sql/planner/LogicalPlanner.java

```diff
@@ -448,6 +453,18 @@ else if (isUsePreferredWritePartitioning(session)) {
             }
         }
 
+        // TODO: Is this too fragile?  Other places we depend on the correspondence between order
+        //  of non-hidden columns and order of symbols, but it feels a little grotty.
```

Comment:
`plan.getFieldMappings()` is guaranteed to match the order of columns in plan.getScope().getRelationType() (i.e., the output "shape" of the query).

The planner is not the right place to match ColumnMetadata to fields, as there is no correspondence between those fields and the order in which the fields appear in the plan.

For instance, given a table `t (a BIGINT, b BIGINT)`, the following query will see the fields in a different order: `INSERT INTO t(b, a) VALUES (1, 10)`

This should be handled during analysis. The analyzer should record which fields ordinals are supposed to be not null. Take a look at `Analysis.JoinUsingAnalysis` and callers for an example of how you might record that.

Reply:
Ok, I'll look there.

Reply:
@martint The two callers of this method in this class fetch the columns names from table metadata. Are they wrong as well?

Reply:
For `CREATE TABLE AS ...` it kind of works out because the metadata is derived from the CREATE TABLE statement. For INSERT INTO, it works because it creates a projection to match the layout of the table metadata. Unfortunately, it relies on `projectNode.getOutputSymbols()`, which is not guaranteed to come out in the desired order -- it just happens to work today.

Reply:
@martint, I believe I've addressed your concern in the latest force push of the commit.  The code no longer expects channel order to match column order, and I added a unit test that exercises your `INSERT INTO t(b, a)` example